overlord/snapstate/handlers.go

```diff
@@ -1978,6 +1978,146 @@ func (m *SnapManager) doPreferAliases(t *state.Task, _ *tomb.Tomb) error {
 	return nil
 }
 
+// changeReadyUpToTask returns whether all other change's tasks are Ready.
+func changeReadyUpToTask(task *state.Task) bool {
+	me := task.ID()
+	change := task.Change()
+	for _, task := range change.Tasks() {
+		if me == task.ID() {
+			// ignore self
+			continue
+		}
+		if !task.Status().Ready() {
+			return false
+		}
+	}
+	return true
+}
+
+// refreshedSnaps returns the instance names of the snaps successfully refreshed
+// in the last batch of refreshes before the given (re-refresh) task.
+//
+// It does this by advancing through the given task's change's tasks, keeping
+// track of the instance names from the first SnapSetup in every lane, stopping
+// when finding the given task, and resetting things when finding a different
+// re-refresh task (that indicates the end of a batch that isn't the given one).
+func refreshedSnaps(reTask *state.Task) []string {
+	// NOTE nothing requires reTask to be a check-rerefresh task, nor even to be in
+	// a refresh-ish change, but it doesn't make much sense to call this otherwise.
+	tid := reTask.ID()
+	laneSnaps := map[int]string{}
+	// change.Tasks() preserves the order tasks were added, otherwise it all falls apart
+	for _, task := range reTask.Change().Tasks() {
+		if task.ID() == tid {
+			// we've reached ourselves; we don't care about anything beyond this
+			break
+		}
+		if task.Kind() == "check-rerefresh" {
+			// we've reached a previous check-rerefresh (but not ourselves).
+			// Only snaps in tasks after this point are of interest.
+			laneSnaps = map[int]string{}
+		}
+		lanes := task.Lanes()
+		if len(lanes) != 1 {
+			// can't happen, really
+			continue
+		}
+		lane := lanes[0]
+		if lane == 0 {
+			// not really a lane
+			continue
+		}
+		if task.Status() != state.DoneStatus {
+			// ignore non-successful lane (1)
+			laneSnaps[lane] = ""
+			continue
+		}
+		if _, ok := laneSnaps[lane]; ok {
+			// ignore lanes we've already seen (including ones explicitly ignored in (1))
```

Comment:
(nitpick) I guess I'm slightly blind, maybe expand the comment `// ignore lanes we've already seen (including ones explicitly ignored in the check above in (1))`